cdap-gateway/src/main/java/co/cask/cdap/gateway/discovery/UserServiceEndpointStrategy.java

```diff
@@ -39,48 +39,77 @@
   private final ProgramId serviceId;
   private final String version;
   private final Random random;
+  private final RouteFallbackStrategy fallbackStrategy;
+  private final RandomEndpointStrategy versionedRandomEndpointStrategy;
 
   public UserServiceEndpointStrategy(ServiceDiscovered serviceDiscovered, RouteStore routeStore, ProgramId serviceId,
-                                     @Nullable String version) {
+                                     RouteFallbackStrategy fallbackStrategy, @Nullable String version) {
     super(serviceDiscovered);
     this.routeStore = routeStore;
     this.serviceId = serviceId;
     this.version = version;
     this.random = ThreadLocalRandom.current();
+    this.fallbackStrategy = fallbackStrategy;
+    this.versionedRandomEndpointStrategy = new RandomEndpointStrategy(
+      new VersionFilteredServiceDiscovered(serviceDiscovered, version));
   }
 
-  public UserServiceEndpointStrategy(ServiceDiscovered serviceDiscovered, RouteStore routeStore, ProgramId serviceId) {
-    this(serviceDiscovered, routeStore, serviceId, null);
+  @VisibleForTesting
+  UserServiceEndpointStrategy(ServiceDiscovered serviceDiscovered, RouteStore routeStore, ProgramId serviceId) {
+    this(serviceDiscovered, routeStore, serviceId, RouteFallbackStrategy.RANDOM, null);
   }
 
   @Nullable
   @Override
   public Discoverable pick() {
-    // If version is not null, then don't fetch any routeConfig.
-    RouteConfig routeConfig = version == null ? routeStore.fetch(serviceId) : null;
+    if (version != null) {
+      return versionedRandomEndpointStrategy.pick();
+    }
+
+    RouteConfig routeConfig = routeStore.fetch(serviceId);
+    boolean routeConfigUnAvailable = (routeConfig == null) || routeConfig.getRoutes().isEmpty();
```

Comment:
I thought about it but all the if conditions need routeConfigUnAvailable. Otherwise it would be if(!hasRouteConfig). I can flip it if you think it will make it more readable.